File: cpp/src/2023/01.cpp
#include "01.h"
#include "../utils/utils.cpp"
#include <array>
#include <iostream>
#include <limits>
#include <map>
#include <string>

using namespace std;

map<string, int> TERMS{{"0", 0},   {"1", 1},     {"2", 2},     {"3", 3},
                       {"4", 4},   {"5", 5},     {"6", 6},     {"7", 7},
                       {"8", 8},   {"9", 9},     {"zero", 0},  {"one", 1},
                       {"two", 2}, {"three", 3}, {"four", 4},  {"five", 5},
                       {"six", 6}, {"seven", 7}, {"eight", 8}, {"nine", 9}};

int main() {
  const array<const string, 10> domain{"0", "1", "2", "3", "4",
                                       "5", "6", "7", "8", "9"};
  cout << "01 a example: (should be 142)" << endl;
  challenge<10>("2023/01a_example.txt", domain);
  cout << "01 a input: (should be 54927)" << endl;
  challenge<10>("2023/01a_input.txt", domain);
  cout << "01 b example: (should be 281)" << endl;

  const array<const string, 20> domain2{
      "0",    "1",    "2",   "3",     "4",     "5",    "6",
      "7",    "8",    "9",   "zero",  "one",   "two",  "three",
      "four", "five", "six", "seven", "eight", "nine",
  };
  challenge<20>("2023/01b_example.txt", domain2);
  cout << "01 b input: (should be 54581)" << endl;
  challenge<20>("2023/01a_input.txt", domain2);
  return 0;
}

template <int SIZE>
void challenge(const string &filePath,
               const array<const string, SIZE> &domain) {
  string fileContent = loadFileContent(filePath);

  if (fileContent.empty()) {
    throw "file " + filePath + " is empty!";
  }
  int sum = algo<SIZE>(fileContent, domain);
  cout << "Sum: " << sum << endl;
}

template <int SIZE>
int algo(const std::string &content, const array<const string, SIZE> &domain) {
  int greedy = -1;
  int lazy = -1;
  int sum = 0;
  for (int i = 0; i < content.length(); i++) {
    char curr = content[i];

    if (curr == '\n') {
      sum += lazy * 10 + greedy;
      greedy = -1;
      lazy = -1;
      continue;
    }

    for (auto term : domain) {
      auto conv = findTermAndConvertToNumber(content, i, term);
      if (conv != -1) {
        greedy = conv;
        if (lazy == -1) {
          lazy = conv;
        }
      }
    }
  }
  return sum;
}

int findTermAndConvertToNumber(const string &content, const int index,
                               const string &term) {
  if (term.length() > content.length() - index)
    return -1;

  for (int i = 0; i < term.length(); i++) {
    if (term[i] != content[i + index])
      return -1;
  }
  return TERMS[term];
}

File: typescript/src/2023/01.ts
import { assert } from "console";
import { loadInput } from "../utils/load-input";


async function solve(file: string, nmap: Record<string, number>) {
  const input = await loadInput(file);
  const lines = input.split("\n").filter(Boolean);
  let sum = 0;
  for (const line of lines) {
    let num = "";
    num += find(line, nmap);
    num += find(line, nmap, "last");
    sum += Number(num);
  }
  return sum;
}


function find(str: string, nmap: Record<string, number>, mode: "first" | "last" = "first") {
  const isFirst = mode === "first";
  if (isFirst) {
    for (let i = 0; i < str.length; i++) {
      for (const n in nmap) {
        if (match(str, i, n)) {
          return nmap[n as keyof typeof nmap];
        }
      }
    }
  } else {
    let found: number | null = null;
    for (let i = 0; i < str.length; i++) {
      for (const n in nmap) {
        if (match(str, i, n)) {
          found = nmap[n as keyof typeof nmap];
        }
      }
    }
    if (found !== null) return found;
  }
  console.log(str)
  throw new Error("Boom! no number found.")
}

function match(str: string, idx: number, word: string) {
  if (word.length > str.length - idx) return false;
  for (let i = idx; i < str.length && (i - idx < word.length); i++) {
    if (str[i] !== word[i - idx]) return false;
  }
  return true;
}

if (require.main === module) {
  const firstNMap = {
    1: 1, 2: 2, 3: 3, 4: 4, 5: 5,
    6: 6, 7: 7, 8: 8, 9: 9,
  };

  const secondNMap = {
    1: 1, 2: 2, 3: 3, 4: 4, 5: 5,
    6: 6, 7: 7, 8: 8, 9: 9,
    one: 1, two: 2, three: 3,
    four: 4, five: 5, six: 6,
    seven: 7, eight: 8, nine: 9
  };
  (async () => {
    const example1 = await solve("2023/01a_example.txt", firstNMap);
    assert(example1 === 142, "example input A should be 142");
    const star1 = await solve("2023/01_input.txt", firstNMap);
    assert(star1 === 54927, "input A one answer should be 54927");
    console.log("first part:", star1);


    const example2 = await solve("2023/01b_example.txt", secondNMap);
    assert(example2 === 281, "example input B should be 281");
    const star2 = await solve("2023/01_input.txt", secondNMap);
    assert(star2 === 54581, "input A one answer should be 54581");
    console.log("second part:", star2);
  })()
}
